src/xrpld/app/tx/detail/LoanSet.cpp

```diff
@@ -0,0 +1,632 @@
+        // Compute defensively. Assure that "tx" cannot be accessed and cause
+        // confusion or miscalculations.
+        return counterSig.isFieldPresent(sfSigners)
+            ? counterSig.getFieldArray(sfSigners).size()
+            : (counterSig.isFieldPresent(sfTxnSignature) ? 1 : 0);
+    }();
+
+    return normalCost + (signerCount * baseFee);
+}
+
+std::vector<OptionaledField<STNumber>> const&
+LoanSet::getValueFields()
+{
+    static std::vector<OptionaledField<STNumber>> const valueFields{
+        ~sfPrincipalRequested,
+        ~sfLoanOriginationFee,
+        ~sfLoanServiceFee,
+        ~sfLatePaymentFee,
+        ~sfClosePaymentFee
+        // Overpayment fee is really a rate. Don't check it here.
+    };
+
+    return valueFields;
+}
+
+static std::uint32_t
+getStartDate(ReadView const& view)
+{
+    return view.info().closeTime.time_since_epoch().count();
+}
+
+TER
+LoanSet::preclaim(PreclaimContext const& ctx)
+{
+    auto const& tx = ctx.tx;
+
+    {
+        // Check for numeric overflow of the schedule before we load any
+        // objects. The Grace Period for the last payment ends at:
+        //     startDate + (paymentInterval * paymentTotal) + gracePeriod.
+        // If that value is larger than "maxTime", the value
+        // overflows, and we kill the transaction.
+        using timeType = decltype(sfNextPaymentDueDate)::type::value_type;
+        static_assert(std::is_same_v<timeType, std::uint32_t>);
+        timeType constexpr maxTime = std::numeric_limits<timeType>::max();
+        static_assert(maxTime == 4'294'967'295);
+
+        auto const timeAvailable = maxTime - getStartDate(ctx.view);
+
+        auto const interval =
+            ctx.tx.at(~sfPaymentInterval).value_or(defaultPaymentInterval);
+        auto const total =
+            ctx.tx.at(~sfPaymentTotal).value_or(defaultPaymentTotal);
+        auto const grace =
+            ctx.tx.at(~sfGracePeriod).value_or(defaultGracePeriod);
+
+        // The grace period can't be larger than the interval. Check it first,
+        // mostly so that unit tests can test that specific case.
+        if (grace > timeAvailable)
+        {
+            JLOG(ctx.j.warn()) << "Grace period exceeds protocol time limit.";
+            return tecKILLED;
+        }
+
+        if (interval > timeAvailable)
+        {
+            JLOG(ctx.j.warn())
+                << "Payment interval exceeds protocol time limit.";
+            return tecKILLED;
+        }
+
+        if (total > timeAvailable)
+        {
+            JLOG(ctx.j.warn()) << "Payment total exceeds protocol time limit.";
+            return tecKILLED;
+        }
+
+        auto const timeLastPayment = timeAvailable - grace;
+
+        if (timeLastPayment / interval < total)
+        {
+            JLOG(ctx.j.warn()) << "Last payment due date, or grace period for "
+                                  "last payment exceeds protocol time limit.";
+            return tecKILLED;
+        }
+    }
```

Comment:
why isn't this block part of the preflight instead and return tem code?

Reply:
> why isn't this block part of the preflight instead and return tem code?

Because of this line
```
auto const timeAvailable = maxTime - getStartDate(ctx.view);
```
The limit can't just be 2^32. It has to be based on the amount of time left between "now" and 2^32. Because `preflight` doesn't have access to the view, it has to be done in here `preclaim`.

I could do a separate check against 2^32 in `preflight`, but that would be redundant, since we'd just have to do it all over again.